### PR TITLE
pipewire: introduce roc-toolkit support

### PR DESCRIPTION
**Pull request by bgamari**
###### Motivation for this change
`roc-streaming` is a library for robust streaming of audio data over (unreliable) networks. `pipewire` has support for it via the `libpipewire-module-roc-{sink,source}` modules. Here I add derivations for the `openfec` and `roc-toolkit` libraries, as well as introduce support for the latter in the `pipewire` derivation.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by bgamari** [comment]
Ccing @jansol

N.B. I have based this on #150999

File: pkgs/development/libraries/openfec/default.nix [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+  nativeBuildInputs = [
+    cmake
+  ];
+
+  configurePhase = ''
+    mkdir build
+    cd build
+    cmake .. -DDEBUG:STRING=OFF
+  '';
```
Comment (Mic92) [inline pkgs/development/libraries/openfec/default.nix:22]:
```suggestion
  cmakeFlags = [ "-DDEBUG:STRING=OFF" ];
```

File: pkgs/development/libraries/openfec/default.nix [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+  installPhase = ''
+    # This is pretty horrible but sadly there is not installation procedure
+    # provided.
+    mkdir -p $dev/include
+    cd ..
+    cp -R src/* $dev/include
+    find $dev/include -type f -a ! -iname '*.h' -delete
```
Comment (Mic92) [inline pkgs/development/libraries/openfec/default.nix:30]:
```suggestion
    find .. -type f -a ! -iname '*.h' -print0 | xargs -0 install -D --target $dev/include
```

File: pkgs/development/libraries/openfec/default.nix [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    mkdir -p $out/lib $dev
+    cd bin/Release
+    install -D -m755 -t $out/lib libopenfec${stdenv.hostPlatform.extensions.sharedLibrary}
```
Comment (Mic92) [inline pkgs/development/libraries/openfec/default.nix:34]:
```suggestion
    install -D -m755 -t $out/lib ../bin/Release/libopenfec${stdenv.hostPlatform.extensions.sharedLibrary}
```

File: pkgs/development/libraries/openfec/default.nix (hunk 2 above)
Comment (Mic92) [reply pkgs/development/libraries/openfec/default.nix:30]: mhm ^ that might not work if you have subdirectories.

File: pkgs/development/libraries/openfec/default.nix (hunk 1 above)
Comment (bgamari) [reply pkgs/development/libraries/openfec/default.nix:22]: Thanks for the review!

Good catch. I forgot that `cmake`'s setup hook already creates `build`.

File: pkgs/development/libraries/openfec/default.nix (hunk 2 above)
Comment (bgamari) [reply pkgs/development/libraries/openfec/default.nix:26]: There indeed are subdirectories, sadly

File: pkgs/development/libraries/openfec/default.nix (hunk 3 above)
Comment (bgamari) [reply pkgs/development/libraries/openfec/default.nix]: I've reworked this. How does it look now?

File: pkgs/development/libraries/audio/roc-toolkit/default.nix [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    [ "--disable-sox"
+      "--disable-tests" ] ++
+    lib.optional (libunwind == null) "--disable-libunwind" ++
+    lib.optional (libpulseaudio == null) "--disable-pulseaudio" ++
+    (if openfec == null
+       then ["--disable-openfec"]
+       else [ "--with-libraries=${openfec}/lib"
+              "--with-openfec-includes=${openfec.dev}/include" ]);
+
+  preConfigure = ''
```
Comment (jansol) [inline pkgs/development/libraries/audio/roc-toolkit/default.nix:43]: Is there a reason why this isn't simply part of `sconsFlags`? Also consider [doing what the godot derivation is doing for scons options](https://github.com/NixOS/nixpkgs/blob/db2ad441370111db38a7d5e52cf74857a0502860/pkgs/development/tools/godot/default.nix#L71).

File: pkgs/development/libraries/openfec/default.nix [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    sha256 = "sha256:0c2lg8afr7lqpzrsi0g44a6h6s7nq4vz7yc9vm2k57ph2y6r86la";
+  };
+
+  outputs = [ "out" "dev" ];
+
+  nativeBuildInputs = [
+    cmake
+  ];
+
+  cmakeFlags = "-DDEBUG:STRING=OFF";
```
Comment (jansol) [inline pkgs/development/libraries/openfec/default.nix:18]: This should be an array of strings.

**Review by jansol — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/audio/roc-toolkit/default.nix [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    libunwind
+    openfec
+    libpulseaudio
+  ];
+
+  sconsFlags =
+    [ "--disable-sox"
+      "--disable-tests" ] ++
+    lib.optional libunwindSupport "--disable-libunwind" ++
+    lib.optional pulseaudioSupport "--disable-pulseaudio" ++
```
Comment (jansol) [inline pkgs/development/libraries/audio/roc-toolkit/default.nix:46]: You probably meant to invert these.
```suggestion
    lib.optional !pulseaudioSupport "--disable-pulseaudio" ++
```

File: pkgs/development/libraries/audio/roc-toolkit/default.nix [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    libuv
+    libunwind
+    openfec
+    libpulseaudio
+  ];
+
+  sconsFlags =
+    [ "--disable-sox"
+      "--disable-tests" ] ++
+    lib.optional libunwindSupport "--disable-libunwind" ++
```
Comment (jansol) [inline pkgs/development/libraries/audio/roc-toolkit/default.nix:45]:
```suggestion
    lib.optional !libunwindSupport "--disable-libunwind" ++
```

**Review by jansol — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/audio/roc-toolkit/default.nix (hunk 4 above)
Comment (jansol) [reply pkgs/development/libraries/audio/roc-toolkit/default.nix:52]: Handling the nix side of things might be cleaner that way, but it is your call. If that method can't handle all options needed here then the benefits are indeed dubious.

File: pkgs/development/libraries/audio/roc-toolkit/default.nix [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    openfec
+    libpulseaudio
+  ];
+
+  sconsFlags =
+    [ "--disable-sox"
+      "--disable-tests" ] ++
+    lib.optional (!libunwindSupport) "--disable-libunwind" ++
+    lib.optional (!pulseaudioSupport) "--disable-pulseaudio" ++
+    (if openfecSupport
```
Comment (jansol) [inline pkgs/development/libraries/audio/roc-toolkit/default.nix:47]: This needs flipping either the condition or the branches.

**Review by jansol — CHANGES_REQUESTED** [review]

**Comment by Mic92** [comment]
@jansol all good?

File: pkgs/development/libraries/audio/roc-toolkit/default.nix (hunk 8 above)
Comment (jansol) [inline pkgs/development/libraries/audio/roc-toolkit/default.nix:47]:
```suggestion
    (if !openfecSupport
```

**Review by jansol — CHANGES_REQUESTED** [review]

**Comment by jansol** [comment]
Haven't tested but looks good with the exception of that openfec flag.

File: pkgs/development/libraries/audio/roc-toolkit/default.nix (hunk 8 above)
Comment (bgamari) [reply pkgs/development/libraries/audio/roc-toolkit/default.nix]: Sigh, indeed I fixed this locally but apparently neglected to push. Sorry about that!

**Review by jansol — APPROVED** [review]

**Comment by bgamari** [comment]
FWIW, I am able to load the `roc-sink` and `roc-source` modules and can stream audio between them.

**Comment by bgamari** [comment]
For the record, this was tested using the backport to 21.11 in https://github.com/bgamari/nixpkgs/tree/wip/rpi-pipewire-21.11.

**Comment by Mic92** [comment]
> For the record, this was tested using the backport to 21.11 in https://github.com/bgamari/nixpkgs/tree/wip/rpi-pipewire-21.11 (which also upgrades `pipewire` to 0.3.42). Perhaps this should be merged to nixos-21.11?

No strong opinion on that but since is a non-trivial change of the package, we should gives this some weeks in the nixos-unstable channel to see if there any regressions.

**Pull request merged**